src/raftstore/store/worker/scan.rs

```diff
@@ -0,0 +1,355 @@
+        start_key: Vec<u8>,
+        end_key: Vec<u8>,
+    },
+    ComputeHash {
+        index: u64,
+        region: Region,
+        snap: Snapshot,
+    },
+    VerifyHash {
+        region_id: u64,
+        index: u64,
+        hash: Vec<u8>,
+    },
+}
+
+impl Task {
+    pub fn split_check(ps: &PeerStorage) -> Task {
+        Task::SplitCheck {
+            region_id: ps.get_region_id(),
+            epoch: ps.get_region().get_region_epoch().clone(),
+            start_key: keys::enc_start_key(&ps.region),
+            end_key: keys::enc_end_key(&ps.region),
+        }
+    }
+
+    pub fn compute_hash(region: Region, index: u64, snap: Snapshot) -> Task {
+        Task::ComputeHash {
+            region: region,
+            index: index,
+            snap: snap,
+        }
+    }
+
+    pub fn verify_hash(region_id: u64, index: u64, hash: Vec<u8>) -> Task {
+        Task::VerifyHash {
+            region_id: region_id,
+            index: index,
+            hash: hash,
+        }
+    }
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            Task::SplitCheck { region_id, .. } => write!(f, "Split Check Task for {}", region_id),
+            Task::ComputeHash { ref region, index, .. } => {
+                write!(f, "Compute Hash Task for {:?} at {}", region, index)
+            }
+            Task::VerifyHash { region_id, index, ref hash } => {
+                write!(f,
+                       "Verify Hash Task for {} at {} [hash: {}]",
+                       region_id,
+                       index,
+                       escape(hash))
+            }
+        }
+    }
+}
+
+pub struct Runner<C: MsgSender> {
+    db: Arc<DB>,
+    ch: C,
+    region_max_size: u64,
+    split_size: u64,
+
+    // region_id -> (index, hash)
+    checksum: HashMap<u64, (u64, Digest)>,
+}
+
+#[inline]
+fn update(sha1: &mut Sha1, v: &[u8]) {
+    let size = v.len();
+    let mut size_arr = [0u8; 8];
+    size_arr.as_mut().encode_u64(size as u64).unwrap();
+    sha1.update(&size_arr);
+    sha1.update(v);
+}
+
+impl<C: MsgSender> Runner<C> {
+    pub fn new(db: Arc<DB>, ch: C, region_max_size: u64, split_size: u64) -> Runner<C> {
+        Runner {
+            db: db,
+            ch: ch,
+            region_max_size: region_max_size,
+            split_size: split_size,
+            checksum: HashMap::new(),
+        }
+    }
+
+    fn compute_hash(&mut self, region: Region, index: u64, snap: Snapshot) {
+        let region_id = region.get_id();
+        info!("[region {}] computing hash at {}", region_id, index);
+        REGION_HASH_COUNTER_VEC.with_label_values(&["compute", "all"]).inc();
+
+        let timer = REGION_HASH_HISTOGRAM.start_timer();
+        let mut sha1 = Sha1::new();
+        let mut cf_names = snap.cf_names();
+        cf_names.sort();
+        let start_key = keys::enc_start_key(&region);
+        let end_key = keys::enc_end_key(&region);
+        for cf in cf_names {
```

Comment:
we can't check raft cf here. The raft log GC is in another thread, we can't guarantee all peers have same raft logs when computing hash.